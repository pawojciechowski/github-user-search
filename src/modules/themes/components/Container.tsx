import styled from 'styled-components';
import { Theme } from '../types';

const Container = styled.div`
  padding-left: ${({ theme }: { theme: Theme }) => `${theme.spacer * 1.5}px`};
  padding-right: ${({ theme }: { theme: Theme }) => `${theme.spacer * 1.5}px`};

  @media screen and (min-width: 768px) {
    max-width: 700px;
    margin: auto;
  }
`;

export default Container;
