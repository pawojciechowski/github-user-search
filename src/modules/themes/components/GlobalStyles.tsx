import { baseFontCSS } from '../styles';
import { createGlobalStyle } from 'styled-components';
import { Theme } from '../types';

export default createGlobalStyle`
  html,
  body {
    height: 100%;
  }

  body {
    ${baseFontCSS}
    -webkit-font-smoothing: antialiased;
    -moz-osx-font-smoothing: grayscale;
    margin: 0;
    padding: 0;
    background: ${({ theme }: { theme: Theme }) => theme.background};
  }

  #root {
    height: 100%;
  }

  img {
    max-width: 100%;
    display: block;
  }
`;
