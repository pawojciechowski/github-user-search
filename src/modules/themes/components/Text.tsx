import styled from 'styled-components';
import { Theme } from '../types';

const Text = styled.p`
  margin: 0;
  padding-bottom: ${({ theme }: { theme: Theme }) => `${theme.spacer}px`};
`;

export default Text;
