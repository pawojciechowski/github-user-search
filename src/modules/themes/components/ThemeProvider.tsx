import React, { PropsWithChildren } from 'react';
import { ThemeProvider as StyledComponentsThemeProvider, createGlobalStyle } from 'styled-components';
import { Theme, ThemeName } from '../types';
import { themes } from '../themes';
import { baseFontCSS } from '../styles';

const GlobalStyle = createGlobalStyle`
  html,
  body {
    height: 100%;
  }

  body {
    ${baseFontCSS}
    -webkit-font-smoothing: antialiased;
    -moz-osx-font-smoothing: grayscale;
    margin: 0;
    padding: 0;
    background: ${({ theme }: { theme: Theme }) => theme.background};
  }

  #root {
    height: 100%;
  }

  img {
    max-width: 100%;
    display: block;
  }
`;

export function ThemeProvider({ children }: PropsWithChildren<{}>) {
  // TODO: Get theme from state
  const themeName: ThemeName = 'default';

  return (
    <StyledComponentsThemeProvider theme={themes[themeName]}>
      <GlobalStyle />
      {children}
    </StyledComponentsThemeProvider>
  );
}

export default ThemeProvider;
