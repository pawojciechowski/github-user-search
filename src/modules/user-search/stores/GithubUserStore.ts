import { types, flow, getEnv, Instance, getSnapshot } from 'mobx-state-tree';
import { IAPIService } from 'modules/api/types';
import { GithubUserResponse, GithubReposResponse } from '../../../config/api/types';
import { APIError } from 'modules/api/errors';

export const GithubRepo = types.model('GithubRepo', {
  id: types.identifierNumber,
  name: types.string,
  htmlUrl: types.string,
  stars: types.number,
});

export const GithubUser = types.model('GithubUser', {
  login: types.identifier,
  name: types.string,
  bio: types.maybeNull(types.string),
  avatarUrl: types.string,
  repos: types.array(GithubRepo),
});

export const GithubUserStore = types
  .model('GithubUserStore', {
    user: types.maybeNull(GithubUser),
    state: types.optional(
      types.union(
        types.literal('initial'),
        types.literal('pending'),
        types.literal('success'),
        types.literal('error'),
        types.literal('notFound')
      ),
      'initial'
    ),
    error: types.maybe(types.string),
    lastSearchUsername: types.maybe(types.string),
  })
  .actions((self) => ({
    fetchUserWithRepos: flow(function* fetchUserWithRepos(username) {
      self.user = null;
      self.lastSearchUsername = username;
      self.state = 'pending';

      const apiService: IAPIService = getEnv(self).apiService;
      const userResponsePromise = apiService.get<GithubUserResponse>(`/users/${username}`);
      const reposResponsePromise = apiService.get<GithubReposResponse>(`/users/${username}/repos`);

      try {
        const [userResponse, reposResponse]: [GithubUserResponse, GithubReposResponse] = yield Promise.all([
          userResponsePromise,
          reposResponsePromise,
        ]);
        self.user = GithubUser.create({
          login: userResponse.login,
          name: userResponse.name || '',
          bio: userResponse.bio,
          avatarUrl: userResponse.avatar_url,
          repos: reposResponse.map((repo) =>
            GithubRepo.create({
              id: repo.id,
              name: repo.name,
              htmlUrl: repo.html_url,
              stars: repo.stargazers_count,
            })
          ),
        });
        self.state = 'success';
      } catch (error) {
        if (error instanceof APIError && error.response?.status === 404) {
          self.state = 'notFound';
          self.user = null;
        } else {
          self.state = 'error';
          self.error = error.message;
          self.user = null;
        }
      }
    }),
  }))
  .views((self) => ({
    get userWithTopRepos() {
      const snapshot = getSnapshot(self);

      if (snapshot.user === null) {
        return null;
      } else {
        return GithubUser.create({
          ...snapshot.user,
          repos: [...snapshot.user.repos].sort((repo1, repo2) => repo2.stars - repo1.stars).slice(0, 3),
        });
      }
    },
    get isLoading() {
      return self.state === 'pending';
    },
  }));

export interface IGithubUserStore extends Instance<typeof GithubUserStore> {}
export interface IGithubUser extends Instance<typeof GithubUser> {}
