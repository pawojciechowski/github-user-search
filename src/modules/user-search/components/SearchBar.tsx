import React, { useState } from 'react';
import { observer, inject } from 'mobx-react';
import { IGithubUserStore } from '../stores/GithubUserStore';
import Button from 'common/button/components/Button';
import Input from 'common/input/components/Input';
import styled from 'styled-components';
import { Theme } from 'modules/themes/types';

const SearchBarContainer = styled.div`
  display: flex;

  ${Input} {
    flex-grow: 1;
    margin-right: ${({ theme }: { theme: Theme }) => `${theme.spacer}px`};
  }
`;

const handleKeyPress = (e: React.KeyboardEvent<HTMLInputElement>, userStore: IGithubUserStore) => {
  if (e.key === 'Enter') {
    userStore.fetchUserWithRepos(e.currentTarget.value);
  }
};

const handleChange = (
  e: React.FormEvent<HTMLInputElement>,
  setInputValue: React.Dispatch<React.SetStateAction<string>>
) => {
  setInputValue(e.currentTarget.value);
};

interface SearchBarProps {
  userStore: IGithubUserStore;
}

export function SearchBar({ userStore }: SearchBarProps) {
  const [value, setInputValue] = useState('');
  return (
    <SearchBarContainer>
      <Input
        type="text"
        value={value}
        onChange={(e: React.FormEvent<HTMLInputElement>) => handleChange(e, setInputValue)}
        onKeyPress={(e: React.KeyboardEvent<HTMLInputElement>) => handleKeyPress(e, userStore)}
        placeholder="Search for users"
      />
      <Button disabled={userStore.isLoading} onClick={() => userStore.fetchUserWithRepos(value)}>
        Search
      </Button>
    </SearchBarContainer>
  );
}

export default inject('userStore')(observer(SearchBar)) as any;
