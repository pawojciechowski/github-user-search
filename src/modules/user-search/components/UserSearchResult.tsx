import React from 'react';
import { observer, inject } from 'mobx-react';
import { IGithubUserStore } from '../stores/GithubUserStore';

interface UserSearchResultProps {
  user: IGithubUserStore
}

export function UserSearchResult({ user }: UserSearchResultProps) {
  return (
    <div>

    </div>
  )
}

export default inject('user')(observer(UserSearchResult)) as any;
