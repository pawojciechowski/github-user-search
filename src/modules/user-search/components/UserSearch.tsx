import React from 'react';
import styled from 'styled-components';
import SearchBar from './SearchBar';
import UserSearchResult from './UserSearchResult';
import Container from 'modules/themes/components/Container';
import { Theme } from 'modules/themes/types';

const UserSearchContainer = styled.div``;
const UserSearchHeader = styled(Container)`
  padding-top: ${({ theme }: { theme: Theme}) => `${theme.spacer * 1.5}px`};
  padding-bottom: ${({ theme }: { theme: Theme}) => `${theme.spacer * 1.5}px`};
  background: ${({ theme }: { theme: Theme}) => theme.foreground};
  box-shadow: ${({ theme }: { theme: Theme}) => theme.containerShadow};
`;
const UserSearchResultContainer = styled(Container)`
  padding-top: ${({ theme }: { theme: Theme}) => `${theme.spacer * 1.5}px`};
  padding-bottom: ${({ theme }: { theme: Theme}) => `${theme.spacer * 1.5}px`};
`;

function UserSearch() {
  return (
    <UserSearchContainer>
      <UserSearchHeader>
        <SearchBar />
      </UserSearchHeader>
      <UserSearchResultContainer>
        <UserSearchResult />
      </UserSearchResultContainer>
    </UserSearchContainer>
  )
}

export default UserSearch;
