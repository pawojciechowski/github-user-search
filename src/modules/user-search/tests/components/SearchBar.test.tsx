import React from 'react';
import { SearchBar } from '../../components/SearchBar';
import { IGithubUserStore } from 'modules/user-search/stores/GithubUserStore';
import { renderWithThemeProvider } from 'config/tests/utils';
import { fireEvent } from '@testing-library/react';

const getTestUserStore = (data: {} = {}) =>
  (({
    fetchUserWithRepos: jest.fn<IGithubUserStore['fetchUserWithRepos'], any[]>(),
    isLoading: false,
    ...data,
  } as unknown) as IGithubUserStore);

const testUsername = 'testUsername';

describe('<SearchBar />', () => {
  it('should render without crashing', () => {
    renderWithThemeProvider(<SearchBar userStore={getTestUserStore()} />);
  });

  it('should match snapshot', () => {
    const { container } = renderWithThemeProvider(<SearchBar userStore={getTestUserStore()} />);
    expect(container.firstChild).toMatchSnapshot();
  });

  it('should properly update value of the input', () => {
    const { container } = renderWithThemeProvider(<SearchBar userStore={getTestUserStore()} />);
    const input = container.querySelector('input')!;

    expect(input).toBeTruthy();

    fireEvent.change(input, {
      target: {
        value: testUsername,
      },
    });

    expect(input.value).toBe(testUsername);
  });

  it('should fetchUserWithRepos on key press (Enter)', () => {
    const userStore = getTestUserStore();
    const fetchUserWithReposMock = userStore.fetchUserWithRepos as jest.MockedFunction<
      IGithubUserStore['fetchUserWithRepos']
    >;
    const { container } = renderWithThemeProvider(<SearchBar userStore={userStore} />);
    const input = container.querySelector('input')!;

    expect(input).toBeTruthy();

    fireEvent.change(input, {
      target: {
        value: testUsername,
      },
    });
    fireEvent.keyPress(input, {
      key: 'Enter',
      keyCode: 13,
    });

    expect(fetchUserWithReposMock.mock.calls.length).toBe(1);
    expect(fetchUserWithReposMock.mock.calls[0][0]).toBe(testUsername);
  });

  it('should fetchUserWithRepos on button click', () => {
    const userStore = getTestUserStore();
    const fetchUserWithReposMock = userStore.fetchUserWithRepos as jest.MockedFunction<
      IGithubUserStore['fetchUserWithRepos']
    >;
    const { container } = renderWithThemeProvider(<SearchBar userStore={userStore} />);
    const input = container.querySelector('input')!;
    const button = container.querySelector('button')!;

    expect(input).toBeTruthy();
    expect(button).toBeTruthy();

    fireEvent.change(input, {
      target: {
        value: testUsername,
      },
    });
    fireEvent.click(button);

    expect(fetchUserWithReposMock.mock.calls.length).toBe(1);
    expect(fetchUserWithReposMock.mock.calls[0][0]).toBe(testUsername);
  });

  it('should disable button when store is in `isLoading` state', () => {
    const { container } = renderWithThemeProvider(<SearchBar userStore={getTestUserStore({ isLoading: true })} />);
    const button = container.querySelector('button')!;

    expect(button).toBeTruthy();
    expect(button.disabled).toBeTruthy();
  });

  it('should disable input when store is in `isLoading` state', () => {
    const { container } = renderWithThemeProvider(<SearchBar userStore={getTestUserStore({ isLoading: true })} />);
    const input = container.querySelector('input')!;

    expect(input).toBeTruthy();
    expect(input.disabled).toBeTruthy();
  });

  it('should not fire fetchUserWithRepos, when isLoading is true - button is disabled', () => {
    const userStore = getTestUserStore({ isLoading: true });
    const fetchUserWithReposMock = userStore.fetchUserWithRepos as jest.MockedFunction<
      IGithubUserStore['fetchUserWithRepos']
    >;
    const { container } = renderWithThemeProvider(<SearchBar userStore={userStore} />);
    const input = container.querySelector('input')!;
    const button = container.querySelector('button')!;

    expect(input).toBeTruthy();
    expect(button).toBeTruthy();

    fireEvent.change(input, {
      target: {
        value: testUsername,
      },
    });
    fireEvent.click(button);

    expect(fetchUserWithReposMock.mock.calls.length).toBe(0);
  });

  it('should not fire fetchUserWithRepos, when isLoading is true - input is disabled', () => {
    const userStore = getTestUserStore({ isLoading: true });
    const fetchUserWithReposMock = userStore.fetchUserWithRepos as jest.MockedFunction<
      IGithubUserStore['fetchUserWithRepos']
    >;
    const { container } = renderWithThemeProvider(<SearchBar userStore={userStore} />);
    const input = container.querySelector('input')!;

    expect(input).toBeTruthy();

    fireEvent.change(input, {
      target: {
        value: testUsername,
      },
    });
    fireEvent.keyPress(input, {
      key: 'Enter',
      keyCode: 13,
    });

    expect(fetchUserWithReposMock.mock.calls.length).toBe(0);
  });
});
