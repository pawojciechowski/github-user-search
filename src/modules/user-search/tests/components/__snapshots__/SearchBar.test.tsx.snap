// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<SearchBar /> should match snapshot 1`] = `
.c4 {
  font-family: test;
  font-size: test;
  line-height: test;
  font-weight: test;
  -webkit-letter-spacing: test;
  -moz-letter-spacing: test;
  -ms-letter-spacing: test;
  letter-spacing: test;
  color: test;
  font-size: 12px;
  line-height: 12px;
  padding: 0.75px;
  background: test;
  color: test;
  border-radius: 8px;
  font-weight: bold;
  white-space: nowrap;
  border: none;
  cursor: pointer;
  -webkit-transition: color 0.15s ease-in-out,background 0.15s ease-in-out;
  transition: color 0.15s ease-in-out,background 0.15s ease-in-out;
}

.c4[disabled] {
  background: testDisabled;
  color: testDisabled;
}

.c3 {
  font-family: test;
  font-size: test;
  line-height: test;
  font-weight: test;
  -webkit-letter-spacing: test;
  -moz-letter-spacing: test;
  -ms-letter-spacing: test;
  letter-spacing: test;
  color: test;
  color: test;
  line-height: 16px;
  padding: 0.625px 0.625px 0.625px 32px;
  background: test;
  border-radius: 8px;
  border: none;
  box-sizing: border-box;
  width: 100%;
}

.c3::-webkit-input-placeholder {
  color: test;
}

.c3::-moz-placeholder {
  color: test;
}

.c3:-ms-input-placeholder {
  color: test;
}

.c3::placeholder {
  color: test;
}

.c2 {
  position: relative;
}

.c2:before {
  content: '';
  position: absolute;
  left: 8px;
  top: 10px;
  width: 16px;
  height: 16px;
  background-image: url(test);
}

.c0 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
}

.c0 .c1 {
  -webkit-box-flex: 1;
  -webkit-flex-grow: 1;
  -ms-flex-positive: 1;
  flex-grow: 1;
  margin-right: 1px;
}

<div
  class="c0"
>
  <div
    class="c1 c2"
  >
    <input
      class="c3"
      placeholder="Search for users"
      type="text"
      value=""
    />
  </div>
  <button
    class="c4"
  >
    Search
  </button>
</div>
`;
